core/ledger/kvledger/snapshot.go

```diff
@@ -0,0 +1,199 @@
+	snapshotMetadataHashFileName = "_snapshot_additional_info.json"
+	jsonFileIndent               = "    "
+)
+
+// snapshotSignableMetadata is used to build a JSON that represents a unique snapshot and
+// can be signed by the peer. Hashsum of the resultant JSON is intended to be used as a single
+// hash of the snapshot, if need be.
+type snapshotSignableMetadata struct {
+	ChannelName        string            `json:"channel_name"`
+	ChannelHeight      uint64            `json:"channel_height"`
+	LastBlockHashInHex string            `json:"last_block_hash"`
+	FilesAndHashes     map[string]string `json:"snapshot_files_raw_hashes"`
+}
+
+type snapshotAdditionalInfo struct {
+	SnapshotHashInHex        string `json:"snapshot_hash"`
+	LastBlockCommitHashInHex string `json:"last_block_commit_hash"`
+}
+
+// generateSnapshot generates a snapshot. This function should be invoked when commit on the kvledger are paused
+// after committing the last block fully and further the commits should not be resumed till this function finishes
+func (l *kvLedger) generateSnapshot() error {
+	snapshotsRootDir := l.snapshotsConfig.SnapshotsRootDir
+	bcInfo, err := l.GetBlockchainInfo()
+	if err != nil {
+		return err
+	}
+	snapshotTempDir, err := ioutil.TempDir(
+		SnapshotsTempDir(snapshotsRootDir),
+		fmt.Sprintf("%s-%d-", l.ledgerID, bcInfo.Height),
+	)
+	if err != nil {
+		return errors.Wrapf(err, "error while creating temp dir [%s]", snapshotTempDir)
+	}
+	newHashFunc := func() (hash.Hash, error) {
+		return l.hashProvider.GetHash(snapshotHashOpts)
+	}
+	txIDsExportSummary, err := l.blockStore.ExportTxIds(snapshotTempDir, newHashFunc)
+	if err != nil {
+		return err
+	}
+	configsHistoryExportSummary, err := l.configHistoryRetriever.ExportConfigHistory(snapshotTempDir, newHashFunc)
+	if err != nil {
+		return err
+	}
+	stateDBExportSummary, err := l.txmgr.ExportPubStateAndPvtStateHashes(snapshotTempDir, newHashFunc)
+	if err != nil {
+		return err
+	}
+
+	if err := l.generateSnapshotMetadataFiles(
+		snapshotTempDir, txIDsExportSummary,
+		configsHistoryExportSummary, stateDBExportSummary,
+	); err != nil {
+		return err
+	}
+	if err := syncDir(snapshotTempDir); err != nil {
+		return err
+	}
+	slgr := SnapshotsFinalDirForLedger(snapshotsRootDir, l.ledgerID)
+	if err := os.MkdirAll(slgr, 0755); err != nil {
+		return errors.Wrapf(err, "error while creating final dir for snapshot:%s", slgr)
+	}
+	if err := syncParentDir(slgr); err != nil {
+		return err
+	}
+	slgrht := SnapshotFinalDirForLedgerHeight(l.snapshotsConfig.SnapshotsRootDir, l.ledgerID, bcInfo.Height)
+	if err := os.Rename(snapshotTempDir, slgrht); err != nil {
+		return errors.Wrapf(err, "error while renaming dir [%s] to [%s]:", snapshotTempDir, slgrht)
+	}
+	if err := syncParentDir(slgrht); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (l *kvLedger) generateSnapshotMetadataFiles(
+	dir string,
+	txIDsExports,
+	configsHistoryExports,
+	stateDBExports map[string][]byte) error {
```

Comment:
These are the export 'summaries', rather than the actual exports, right? May want to clarify that in the variable names.

Reply:
yes, these are summaries. Changed the variable names now.